src/ir/types/type.h

```diff
@@ -31,13 +46,13 @@ class Type {
   const TypeBase& type_base() const { return *type_.get(); }
 
   friend class TypeFactory;
+
  private:
-  Type(std::unique_ptr<TypeBase> type) : type_(std::move(type)) {}
+  Type(std::unique_ptr<TypeBase> type) : type_(type.release()) {}
```

Comment:
I'm not sure we should accept a `std::unique_ptr` as a parameter here. It seems like some confusing mixing of proofs. Perhaps `Type` should just accept a `TypeBase` value, and then the constructor for `type_` can just be

```C++
type_(new TypeBase(std::move(type))
```

Reply:
This is is a `unique_ptr` where we explicitly transfer ownership to the object? I agree this would be a problem if it is a `shared_ptr`. Moreover, https://en.cppreference.com/w/cpp/memory/unique_ptr/release explicitly releases ownership?

Also, this is a quick incremental change with having to change the places where `Type` is constructed.

Reply:
ok, sounds good then.

Reply:
On second thought, even though this is a `unique_ptr`, it may not be a good idea to mix these two. Things will be OK as long as the pointer cleanly changes ownership from `unique_ptr` to `intrusive_ptr`. However, if these ever intertwine, it won't be pretty!

Nevertheless, let us get this as it is safe _currently_. We should fix this soon. I have filed #406.